listenbrainz/webserver/static/js/src/stats/UserEntity.tsx

```diff
@@ -0,0 +1,449 @@
+        entityCount = data.payload.total_artist_count;
+      } else if (entity === "release") {
+        data = data as UserReleasesResponse;
+        maxListens = data.payload.releases[0].listen_count;
+        totalPages = Math.ceil(
+          data.payload.total_release_count / this.ROWS_PER_PAGE
+        );
+        entityCount = data.payload.total_release_count;
+      }
+
+      data = await this.getData(page, newRange, entity);
+      await new Promise((resolve) =>
+        this.setState(
+          {
+            data: this.processData(data, page),
+            range: newRange,
+            currPage: page,
+            startDate: new Date(data.payload.from_ts * 1000),
+            totalPages,
+            maxListens,
+            entityCount,
+          },
+          resolve
+        )
+      );
+    } catch (error) {
+      this.handleError(error);
+    }
+  };
+
+  changeEntity = async (newEntity: Entity): Promise<void> => {
+    const { range } = this.state;
+    const { user } = this.props;
+
+    const page = 1;
+    try {
+      let data = await this.APIService.getUserEntity(
+        user.name,
+        newEntity,
+        range,
+        undefined,
+        1
+      );
+
+      let maxListens = 0;
+      let totalPages = 0;
+      let entityCount = 0;
+
+      if (newEntity === "artist") {
+        data = data as UserArtistsResponse;
+        maxListens = data.payload.artists[0].listen_count;
+        totalPages = Math.ceil(
+          data.payload.total_artist_count / this.ROWS_PER_PAGE
+        );
+        entityCount = data.payload.total_artist_count;
+      } else if (newEntity === "release") {
+        data = data as UserReleasesResponse;
+        maxListens = data.payload.releases[0].listen_count;
+        totalPages = Math.ceil(
+          data.payload.total_release_count / this.ROWS_PER_PAGE
+        );
+        entityCount = data.payload.total_release_count;
+      }
+
+      data = await this.getData(page, range, newEntity);
+      this.setState({
+        data: this.processData(data, page, newEntity),
+        entity: newEntity,
+        currPage: page,
+        startDate: new Date(data.payload.from_ts * 1000),
+        totalPages,
+        maxListens,
+        entityCount,
+      });
+    } catch (error) {
+      this.handleError(error);
+    }
+  };
+
+  getData = async (
+    page: number,
+    range: UserEntityAPIRange,
+    entity: Entity
+  ): Promise<UserArtistsResponse | UserReleasesResponse> => {
+    const { user } = this.props;
+    const offset = (page - 1) * this.ROWS_PER_PAGE;
+
+    const data = await this.APIService.getUserEntity(
+      user.name,
+      entity,
+      range,
+      offset,
+      this.ROWS_PER_PAGE
+    );
+    return data;
+  };
+
+  processData = (
+    data: UserArtistsResponse | UserReleasesResponse,
+    page: number,
+    entity?: Entity
+  ): UserEntityData => {
+    if (!entity) {
+      // eslint-disable-next-line no-param-reassign
+      ({ entity } = this.state);
```

Comment:
what does this do?

Reply:
Wait, I guess I understand what it does, do we need the parameter from the function at all? Why not just always use the state?

Reply:
I wonder if this call will be an issue: https://github.com/metabrainz/listenbrainz-server/pull/879/files#diff-1812b9221be0ce4e18b25519f801eabcR190
You'd probably have to set the newEntity before calling processData in a callback.

Reply:
Yes, we need the `newEntity` to be set before calling `processData` in `setState` as MonkeyDo said.